File: android/app/src/main/java/jp/touya/app/domain/StoryRunner.kt
package jp.touya.app.domain

import jp.touya.app.data.ChatMessage
import jp.touya.app.data.StoryBeat
import jp.touya.app.data.StoryChoice

/**
 * Mirror of src/lib/story-runner.ts. Bubble ids are deterministic so a resumed
 * chapter never double-posts. The runner never calls the LLM; only a `free`
 * beat goes through the normal chat send.
 */

private val OPENING_IDS = Regex("^(greeting|welcome-)")

fun storyBubbleId(chapterId: String, beatId: String, index: Any): String = "story-$chapterId-$beatId-$index"

/** Narration caption (if any) followed by one assistant bubble per `text` element. */
fun storyBeatBubbles(chapterId: String, beat: StoryBeat): List<ChatMessage> = buildList {
    beat.narration?.trim()?.takeIf { it.isNotEmpty() }?.let {
        add(ChatMessage("assistant", it, id = storyBubbleId(chapterId, beat.id, "n"), narration = true))
    }
    beat.text.forEachIndexed { index, text ->
        val content = text.trim()
        if (content.isNotEmpty()) add(ChatMessage("assistant", content, id = storyBubbleId(chapterId, beat.id, index)))
    }
}

fun choiceBubble(chapterId: String, beatId: String, choice: StoryChoice): ChatMessage =
    ChatMessage("user", (choice.userText ?: choice.label).trim(), id = "choice-$chapterId-$beatId")

/** Append only bubbles whose id is not already present. */
fun appendUnique(messages: List<ChatMessage>, bubbles: List<ChatMessage>): List<ChatMessage> {
    val seen = messages.map { it.id }.toSet()
    val fresh = bubbles.filter { it.id !in seen }
    return if (fresh.isEmpty()) messages else messages + fresh
}

/** Drop the greeting / welcome-back bubbles: the chapter speaks instead. Saved chat stays. */
fun stripOpening(messages: List<ChatMessage>): List<ChatMessage> =
    messages.filterNot { OPENING_IDS.containsMatchIn(it.id) }

/** Composer is hidden while a non-free beat waits for a tap. */
fun composerHidden(beat: StoryBeat?): Boolean = beat != null && beat.kind != "free"

File: src/lib/story-runner.ts
import type { UiMessage } from "@/components/message-bubble";
import type { StoryBeatPublic, StoryChapterPublic, StoryChoice, StoryScriptPublic } from "./story-types";

/**
 * Client-side pure helpers for the story runner (Web and, by mirror, Android).
 * Bubble ids are deterministic so re-renders and resumes never double-post.
 */

export const OPENING_IDS = /^(greeting|welcome-)/;

export function findChapterPublic(script: StoryScriptPublic | null | undefined, chapterId: string | null | undefined): StoryChapterPublic | null {
  if (!script || !chapterId) return null;
  return script.chapters.find((chapter) => chapter.id === chapterId) ?? null;
}

export function findBeatPublic(
  script: StoryScriptPublic | null | undefined,
  chapterId: string | null | undefined,
  beatId: string | null | undefined,
): StoryBeatPublic | null {
  const chapter = findChapterPublic(script, chapterId);
  if (!chapter || !beatId) return null;
  return chapter.beats.find((beat) => beat.id === beatId) ?? null;
}

export function storyBubbleId(chapterId: string, beatId: string, index: number | "n"): string {
  return `story-${chapterId}-${beatId}-${index}`;
}

/** Narration caption (if any) followed by one assistant bubble per `text` element. */
export function storyBeatBubbles(chapterId: string, beat: StoryBeatPublic): UiMessage[] {
  const out: UiMessage[] = [];
  if (beat.narration?.trim()) {
    out.push({ id: storyBubbleId(chapterId, beat.id, "n"), role: "assistant", content: beat.narration.trim(), narration: true });
  }
  beat.text.forEach((text, index) => {
    const content = text.trim();
    if (content) out.push({ id: storyBubbleId(chapterId, beat.id, index), role: "assistant", content });
  });
  return out;
}

export function choiceBubble(chapterId: string, beatId: string, choice: Pick<StoryChoice, "label" | "userText">): UiMessage {
  return {
    id: `choice-${chapterId}-${beatId}`,
    role: "user",
    content: (choice.userText ?? choice.label).trim(),
  };
}

/** Append only bubbles whose id is not already present. */
export function appendUnique(messages: UiMessage[], bubbles: UiMessage[]): UiMessage[] {
  const seen = new Set(messages.map((message) => message.id));
  const fresh = bubbles.filter((bubble) => !seen.has(bubble.id));
  return fresh.length === 0 ? messages : [...messages, ...fresh];
}

/** Drop the SSR greeting / welcome-back bubbles: the chapter speaks instead. Saved chat stays. */
export function stripOpening(messages: UiMessage[]): UiMessage[] {
  const next = messages.filter((message) => !OPENING_IDS.test(message.id));
  return next.length === messages.length ? messages : next;
}

/** Composer is hidden while a non-free beat waits for a tap. */
export function composerHidden(beat: StoryBeatPublic | null): boolean {
  return beat !== null && beat.kind !== "free";
}
